### DISCUSSION TITLE
Reproducibility

### DISCUSSION THREAD

**Discussion by pibion** [cssi-pi-community/2022-meeting#111] [Special topics breakouts (Tuesday afternoon)]
Welcome to the reproducibility discussion!

**Comment by pibion** [comment]
Suggested topic: what is the status of reproducibility?

**Reply by pibion** [reply]
Representative from the fluid dynamics community.  Reviewers used to have to try to reproduce results.  But now MFL's cutoff for reviewers is two weeks and it's impossible to reproduce in that time frame!

**Reply by pibion** [reply]
Would it be helpful to have both a content reviewer and a data analysis reviewer?  

It's very hard to get even a single reviewer!

Point: journals will usually give you extra time.

But we do think this is a really good challenge.

**Reply by pibion** [reply]
Economics journals are also beginning to enforce this.

**Reply by horsburgh** [reply]
Some journals are now requiring artifact review.

**Reply by pkgw** [reply]
Other journals have "data editors" that work with authors to increase reproducibility and reusability of results.

**Reply by ludaesch** [reply]
Yes, here are some references:
- Report by the AEA Data Editor, Lars Vilhuber, AEA PAPERS AND PROCEEDINGS, VOL. 111, MAY 2021 (pp. 808-17) [https://www.aeaweb.org/articles/pdf/doi/10.1257/pandp.111.808](https://www.aeaweb.org/articles/pdf/doi/10.1257/pandp.111.808)
- Report by the AEA Data Editor, Lars Vilhuber, AEA PAPERS AND PROCEEDINGS, VOL. 112, MAY 2022 (pp. 813-23) [https://www.aeaweb.org/articles/pdf/doi/10.1257/pandp.112.813](https://www.aeaweb.org/articles/pdf/doi/10.1257/pandp.112.813)

**Comment by pibion** [comment]
"Reproducibility" begs three questions:

1. What do we mean by it?
2. Who provides it?
3. Who uses it?
4. Where does provenance come it?

**Reply by ludaesch** [reply]
Here are some references (which contain further references, of course :-) 

- Committee on Reproducibility and Replicability in Science. **Reproducibility and Replicability in Science**. The National Academies Press, Washington, DC, 2019. [https://www.nationalacademies.org/our-work/reproducibility-and-replicability-in-science](https://www.nationalacademies.org/our-work/reproducibility-and-replicability-in-science)

- T. McPhillips, C. Willis, M. R. Gryk, S. Nuñez-Corrales and B. Ludäscher, "**Reproducibility by Other Means: Transparent Research Objects**," 2019 15th International Conference on eScience (eScience), 2019, pp. 502-509, [https://ieeexplore.ieee.org/document/9041774](https://ieeexplore.ieee.org/document/9041774)
 [https://api.semanticscholar.org/CorpusID:214595230](https://api.semanticscholar.org/CorpusID:214595230) 

- Willis, C., & Stodden, V. (2020). **Trust but Verify: How to Leverage Policies, Workflows, and Infrastructure to Ensure Computational Reproducibility in Publication**. Issue 2.4, Fall 2020. [https://api.semanticscholar.org/CorpusID:230602904](https://api.semanticscholar.org/CorpusID:230602904)

- Vilhuber, Lars. "**Reproducibility and replicability in economics**." (2020). [https://hdsr.mitpress.mit.edu/pub/fgpmpj1l/release/5](https://hdsr.mitpress.mit.edu/pub/fgpmpj1l/release/5)

**Reply by ludaesch** [reply]
Someone just mentioned "auditability" rather then reproducibility/replicability/rerunnability.
Indeed, we can view **trustworthiness** and **transparency** as more fundamental than re-executability. See the above reference on "Reproducibility by other means.." as well as the following workshop: _T7: Workshop on Provenance for Transparent Research_
https://iitdbgroup.github.io/ProvenanceWeek2021/t7.html  (also: https://t7.tmcphillips.dev/)

**Comment by pibion** [comment]
# What is Reproducibility?

**Reply by pibion** [reply]
Being able to replicate a result

**Reply by pibion** [reply]
Numerical reproducibility can be nearly impossible.  
Also, reproducing performance is difficult because it's so hardware dependent.

What's more important is the reproducibility of a scientific result.  If an analysis results in a picture of a black hole, others in the fields should be able to start with the same data, use the same workflow, and also see a black hole with a similar statistical significance.

**Reply by pibion** [reply]
There is a paper by Victoria Stodden et al. discussing community efforts to define reproducibility (?)

**Reply by horsburgh** [reply]
Ideas about reproducibility from the hydrology modeling domain:
https://doi.org/10.1016/j.envsoft.2020.104753
https://doi.org/10.1016/j.envsoft.2020.104888

**Reply by pibion** [reply]
Verification and validation is another part that often comes before working on reproducibility.

**Reply by pibion** [reply]
Benchmarking is an important aspect of creating new methods/algorithms.

Thought from Amy: benchmarking feels like a more useful target than numerical reproducibility, which as mentioned above is nearly impossible.

**Reply by pibion** [reply]
https://github.com/cssi-pi-community/2022-meeting/discussions/111#discussioncomment-3236806

**Reply by horsburgh** [reply]
Can't have reproducibility without accessibility. And, you need directions for how to use the accessible artifacts.

**Reply by pkgw** [reply]
Pretty clear that "reproducibility" can be defined in multiple ways. And on the other end, there are a variety of different reasons that we want scientific works to be reproducible. Different reasons potentially link more closely to different definitions?

**Comment by horsburgh** [comment]
## GitHub has its issues 
* Repositories come and go
* Having a repo isn't a standard thing for people to do
* Not everyone freezes a release that matches results describe in a paper, so things can change over time

**Reply by horsburgh** [reply]
Often times, the code is provided, but not all of the instructions are provided. And, often the input data may not be provided. People may say that a ML algorithm is trained using publicly available data, but the details are often not documented to the point that you could do exactly what people did.

**Reply by horsburgh** [reply]
Along with the code, the workflow needs to be captured. What are the steps needed to run the code to get the results? This may be an entire workflow that is much more than just the code.

**Reply by pkgw** [reply]
Researchers often have a naive view of what steps are sufficient to promote reproducibility. Conda "environment" files are increasingly popular, but often they aren't even usable *months* after creation — modern package ecosystems such as NPM/Conda/PyPI move very quickly.

**Comment by pibion** [comment]
# Documentation

**Reply by pibion** [reply]
Documentation can be a great place for people working on reproducing to collaborate with domain scientists.  It's starting to be possible to publish this work!  There are starting to be Journals focused on reproducing results.

**Reply by horsburgh** [reply]
Human readability is important. There are technological solutions to reproducibility, but one argument is for documentation for a human to interpret so that a person could figure it out. 

The counter-argument to this is that it seems unscalable and potentially unreliable.

**Reply by horsburgh** [reply]
There is a strong argument for documentation. Having something that is human readable and something that is machine readable are not necessarily mutually exclusive. They can be done together.

**Reply by horsburgh** [reply]
Benchmark testing is also critical to establishing trust in new methods.

**Comment by dtarb** [comment]
This national academy report gives definitions of reproducibility
National Academies of Sciences, Engineering and Medicine, (2019), Reproducibility and Replicability in Science, The National Academies Press, Washington, DC, 218 p, http://doi.org/10.17226/25303.

**Reply by pkgw** [reply]
ACM also has a set of definitions, which may or may not agree with NASEM.

**Comment by pibion** [comment]
# Requirements for Reproducibility

**Reply by pibion** [reply]
* Access to the code
* Understanding the code

**Reply by horsburgh** [reply]
Access to inputs - e.g., forcing data

**Reply by horsburgh** [reply]
Access to an environment within which the code can be run. This includes:
* Base language and version
* Dependencies, which may also have specific versions

**Reply by pibion** [reply]
An environment where the code can compile!

**Reply by pkgw** [reply]
Returning to the concept that there are various ways to define "reproducibility": we can envision journals marking articles with a variety of "badges" based on whether they meet one or more criteria contributing to reproducibility, i.e. "all software presented in the article is open-source and archived with a DOI". May be more tractable for both authors and publishers than coming up with one umbrella definition of "reproducibility". This approach doesn't need to be limited specifically to traditional journal publishing.

**Comment by dtarb** [comment]
This paper describes work towards reproducibility in Hydrology and Environmental Modeling
Choi, Y.-D., J. L. Goodall, J. M. Sadler, A. M. Castronova, A. Bennett, Z. Li, B. Nijssen, S. Wang, M. P. Clark, D. P. Ames, J. S. Horsburgh, H. Yi, C. Bandaragoda, M. Seul, R. Hooper and D. G. Tarboton, (2021), "Toward Open and Reproducible Environmental Modeling by Integrating Online Data Repositories, Computational Environments, and Model Application Programming Interfaces," Environmental Modelling & Software, 135: 104888, https://doi.org/10.1016/j.envsoft.2020.104888.

**Comment by league** [comment]
Ambiguity and "bit rot" of instructions to build and run, gather dependencies, etc. Sometimes even low-order digit differences can have bigger impact downstream.

**Reply by league** [reply]
Are these getting better, or worse!? A lot of whiz-bang packaging systems evolve too quickly. An old-school C makefile from 15 years ago *might* still work, but an npm/pip freeze file from 6 months ago can have breakage.

**Reply by league** [reply]
Potential solution (or at least a tool) in Nix and Guix strategy: bit-reproducible, build is in a pure "sealed" environment.

**Reply by league** [reply]
But also the *social practice* side of it: when designing, building, documenting… need to keep in mind downstream usability. It's extra work. It's harder than simply getting it to work once… so need incentives, expectations, returns on that work.

**Reply by league** [reply]
Even with fancy tools (maybe *especially* with fancy tools), still need to value *HUMAN* readability/understandability... cannot neglect that.

**Reply by league** [reply]
Counterpoint: better have technical repro: push button, get result, time after time.

**Reply by league** [reply]
Perhaps time scales are the difference: push-button nix/guix technical repro can be good for X years, bot HUMAN readability/understandability may reach 5X or 10X years. Also: *replicability* vs reproducibility.

**Reply by league** [reply]
… vs auditability?

**Comment by league** [comment]
Often even when code is provided/available, all the models/parameters/inputs may not be fully specified. Especially problematic in building ML models, e.g.

**Reply by league** [reply]
Paper may say "we trained it on XYZ publicly-available data sets, code is on github" etc., but it's insufficient to reproduce.

**Reply by league** [reply]
Size of data sets and other inputs can be prohibitive.

**Reply by league** [reply]
Reproducibility needs not just inputs and code, but *workflow* – includes parameters, number of cycles, etc.

**Reply by pkgw** [reply]
Certain workflows might depend on network services, making reproduction fragile (at best) or completely impossible (at worst). As software stacks get more and more complex, it is easier for network dependencies (and similar things that make reproducibility harder) to sneak in even if you're trying to avoid them.

**Comment by horsburgh** [comment]
Developing something so that is reproducible is "extra work" compared to just developing something to answer a problem. What are the incentives to do the social practice that makes things reproducible?

**Comment by league** [comment]
Code verification: 10% of output/effort to do the job itself, 90% to verify that it actually works or does what it says.

**Comment by league** [comment]
Issue of stacks frameworks. Containers/nix/etc can help keep it all in place, but still have a verification burden.

**Comment by league** [comment]
Exists ACM definitions of replicability, reproducibility. There are different levels: artifact available, up to "full reproducibility" of work and intermediate levels that may be interesting. But also these terms can have different definitions/senses in different fields or organizations. Exists National Academy of Sciences report as well. (Links welcome, haha)

**Reply by league** [reply]
Also consider goal of reproducibility... for discovery. For learning.

**Reply by league** [reply]
"Badging" system to represent these intermediate results. Not all or nothing.

**Comment by league** [comment]
The CI/CSSI community should collectively have/develop a position on these issues, and various dimensions of repro. Can get at issues of ethical/appropriate scientific conduct esp in biomedical apps, human subjects, etc.

**Reply by league** [reply]
Maintainers/scientists **responding to emails** can be part of reproducibility! Also unique identifiers important.

**Reply by league** [reply]
A common white lie in research papers: "code available upon request".

**Reply by league** [reply]
Journals need to crack down on such lies. Code must be in trusted repository.

**Reply by league** [reply]
There exist journals that *will* publish/collect open-source stuff without requiring novelty...? JOSS?

**Reply by league** [reply]
Our role as reviewers, role models for sci communities, is to ensure that code is documented, published in appropriate software journal.

**Reply by pkgw** [reply]
JOSS lives here: https://joss.theoj.org/
